### PR TITLE
StringUtil: Fix possible bad free

### PR DESCRIPTION
**Pull request by neomonkey**
This patch fixes strtoul functionts return test. The function can return also EINVAL which was not taken in account.

The second bug is a memory error and possibly a double free bug in non-Windows machines.

### CODE REVIEW

File: Source/Core/Common/StringUtil.cpp [hunk 1]
```diff
@@ -168,13 +168,19 @@ std::string StringFromFormatV(const char* format, va_list args)
...
+      buf = nullptr;
+  }
+
 #if !defined(ANDROID) && !defined(__HAIKU__) && !defined(__OpenBSD__)
   uselocale(previousLocale);
 #endif
 
   std::string temp = buf;
-  free(buf);
+  if (buf != nullptr)
```
Comment (sepalani) [inline Source/Core/Common/StringUtil.cpp]: The if is useless since you're allowed to free nullptr according to ```man free(3)```.
> If ptr is NULL, no operation is performed.

File: Source/Core/Common/StringUtil.cpp [hunk 2]
```diff
@@ -77,13 +77,10 @@ bool AsciiToHex(const std::string& _szValue, u32& result)
   // Set errno to a good state.
   errno = 0;
 
-  char* endptr = nullptr;
-  const u32 value = strtoul(_szValue.c_str(), &endptr, 16);
+  const u32 value = strtoul(_szValue.c_str(), nullptr, 16);
 
-  if (!endptr || *endptr)
```
Comment (sepalani) [inline Source/Core/Common/StringUtil.cpp]: I assume the function is supposed to fail if the value is partially converted. So these checks shouldn't be removed.

File: Source/Core/Common/StringUtil.cpp [hunk 3]
```diff
@@ -77,13 +77,10 @@ bool AsciiToHex(const std::string& _szValue, u32& result)
...
-    return false;
-
-  if (errno == ERANGE)
+  // Fail if errno EINVAL or ERANGE
```
Comment (sepalani) [inline Source/Core/Common/StringUtil.cpp]: According to the ```man strtoul(3)```:
> **EINVAL**
(not in C99) The given base contains an unsupported value.

This should never happen in this case because base 16 is supported.

Comment (neomonkey) [inline Source/Core/Common/StringUtil.cpp]: According to the man strtoul(3):

ERANGE

> 
> The resulting value was out of range.
> The implementation may also set errno to EINVAL in case no conversion was performed (no digits seen, and 0 returned).

File: Source/Core/Common/StringUtil.cpp (hunk 1 above)
Comment (neomonkey) [inline Source/Core/Common/StringUtil.cpp]: This seems to be true

File: Source/Core/Common/StringUtil.cpp (hunk 3 above)
Comment (sepalani) [inline Source/Core/Common/StringUtil.cpp]: That's true in C, not in C++ according to both [en.cppreference.com](http://en.cppreference.com/w/c/string/byte/strtoul) and [cplusplus](http://www.cplusplus.com/reference/cstdlib/strtoul/). The only errno value mentioned is ERANGE. I'm not against changing the if condition but the comment seems irrelevant. Do you have any concrete examples?

Comment (neomonkey) [inline Source/Core/Common/StringUtil.cpp]: I checked those sources also and because there are two different version. So I went to check the standard. Indeed it is not saying anything about EINVAL so I need to change my source :D.

> Source Open STD page 344 -->

File: Source/Core/Common/StringUtil.cpp (hunk 2 above)
Comment (neomonkey) [inline Source/Core/Common/StringUtil.cpp]: According to the  `Open STD page 345`

If the endptr is NULL, then nothing is stored there. So is this check for future purposes; if the nullptr changes in the future to something else?

Comment (sepalani) [inline Source/Core/Common/StringUtil.cpp]: The ```*endptr``` value should be a nul byte if the whole string has been translated. That's why you shouldn't pass nullptr if you want to check strtoul on the whole string.

**Comment by sepalani** [comment]
@neomonkey This PR should probably be squashed into a single commit where only the ```vasprintf``` buffer fix is kept as your other commits seems irrelevant.

**Comment by shuffle2** [comment]
@neomonkey please update the PR. :)

File: Source/Core/Common/StringUtil.cpp [hunk 4]
```diff
@@ -78,13 +78,10 @@ bool AsciiToHex(const std::string& _szValue, u32& result)
...
```
Comment (shuffle2) [inline Source/Core/Common/StringUtil.cpp]: I'm not really a fan of superfluous comments, which I think this is.

**Comment by neomonkey** [comment]
Sorry took a while.. Almost forgot this.

**Comment by spycrab** [comment]
@dolphin-emu-bot rebuild

**Comment by BhaaLseN** [comment]
Please [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) instead of merging. For your case, it should be enough to run `git rebase upstream/master` (preferably after `git fetch upstream`)

**Comment by leoetlino** [comment]
I've rebased your commit. Thanks for your contribution.
